### ISSUE TITLE
Troubles in application

### ISSUE THREAD

**Issue by jkarolczak** [zhouhaoyi/Informer2020#208]
I would to make use of your work in a real-world application. Your paper looks very promising, but unfortunately, I have some problems with implementation. 

Before I proceed further I would like to inform you, that I have built Informer2020 as a .whl package (I did it using non-public Azure DevOps workspace, but for discussion purposes, I uploaded it on GitHub: [https://github.com/jkarolczak/informer-whl](https://github.com/jkarolczak/informer-whl)). I removed from the package files and directories like img, scripts, exp, main_informer.py which seems not to be useful in a real-world application. If you will I can put this file into the PIP indexer. Please notice, that you are listed as the authors.

I use Jupiter Notebook for experiments. I start with installation and necessary imports:

```python
!pip install informer2020-0.0.1-py3-none-any.whl
import informer2020
from informer2020.models.model import Informer
from informer2020.data.data_loader import Dataset_Custom
import torch
import pandas as pd
import matplotlib.pyplot as plt
from matplotlib.pyplot import figure
```

Later on, I define necessary parameters and instantiate object basing on Informer class. I follow your naming convention.

```python
enc_in = 1
dec_in = 1
pred_len = 24
c_out = 1
seq_len = 96
label_len = 48
out_len = 24
size = (seq_len, label_len, pred_len)

model = Informer(enc_in, dec_in, c_out, seq_len, label_len, out_len)
```
The code above contains the first crucial part - setting enc_in and/or dec_in values greater than one returns an error indicating, that model script cannot execute because of mismatching dimensions.

Later on, I prepare data, to make it easier I use file from paper
```python
train_dataset = Dataset_Custom('data/', size=size, data_path='ETTH1-train.csv', freq='h', flag='train', target='OT')
train_data_loader = torch.utils.data.DataLoader(train_dataset, batch_size=32)
```

I perform training with quite standard training loop.
```python
epochs = 30
lr = 0.0001
loss_fn = torch.nn.MSELoss()
optimizer = torch.optim.Adam(model.parameters(), lr)
for e in range(epochs):
    for b, batch in enumerate(train_data_loader):
        batch_x, batch_y, batch_x_mark, batch_y_mark = batch
        dec_inp = torch.zeros([batch_y.shape[0], pred_len, batch_y.shape[-1]]).float()
        dec_inp = torch.cat([batch_y[:,:label_len,:], dec_inp], dim=1).float()
        y_hat = model(batch_x.float(), batch_x_mark.float(), dec_inp.float(), batch_y_mark.float())
        batch_y = batch_y[:,-pred_len:, 0:]
        loss = loss_fn(y_hat, batch_y.float())
        optimizer.zero_grad()
        loss.backward(retain_graph=True)
        optimizer.step()
        print('epoch: {:8} | batch: {:8} | loss:{:14.4f}'.format(e, b, loss.item()), end='\r')
```

Results of such a training are very unsatisfying as MSE is expressed in full tenths. Empirical judgment of the last batch of the training set is also quite disappointing. 

![image](https://user-images.githubusercontent.com/59308844/128878535-b44a58a9-e5f9-4d5f-bfbd-02a336671de6.png)

Below I show how I obtain the plot. I know it's quite a dumb way, but I will to fix previous issues before spending time on making it more elegant and easier to apply in real-world applications.

```python3
test_dataset = Dataset_Custom('data/', size=size, data_path='ETTH1-train.csv', freq='h', flag='test', target='OT')
test_data_loader = torch.utils.data.DataLoader(test_dataset, batch_size=len(test_dataset), drop_last=False)
_ = model.eval()

for batch in test_data_loader:
    batch_x, batch_y, batch_x_mark, batch_y_mark = batch    
    dec_inp = torch.zeros([batch_y.shape[0], pred_len, batch_y.shape[-1]]).float()
    dec_inp = torch.cat([batch_y[:,:label_len,:], dec_inp], dim=1).float()
    batch_y = batch_y[:,-pred_len:, 0:]
    y_hat = model(batch_x.float(), batch_x_mark.float(), dec_inp.float(), batch_y_mark.float())
test_ds = pd.read_csv('data/ETTH1-train.csv')
y = test_ds['OT'][-pred_len:]
y_hat = train_dataset.inverse_transform(torch.flatten(y_hat[-1]).detach().numpy())
fig = figure(figsize=(16, 9))
x = test_ds['date'][-pred_len:]
plt.plot(x, y)
plt.plot(x, y_hat)
```

Could you give me some help and guidance in repairing the code? I will be very glad for your advice. Please, also tell me whether you will or not to index your package to PyPI

**Comment by jkarolczak** [comment]
I found out, that I misused `Dataset_Custom` constructor and after passing `features='MS'` I can set `enc_in` and `dec_in` to `7` without causing any error. Other remarks/issues are still valid

**Comment by jkarolczak** [comment]
I've found that I'm supposed to use `batch_y[:,:, -1]` instead of `batch_y` in the loss computation to avoid mismatching dimensions and take loss only in terms of `OT`. This error wasn't thrown by the interpreter as PyTorch was broadcasting `y_hat` to fit the `batch_y`. Nevertheless, I haven't noticed any improvement in predictions quality.

**Comment by cookieminions** [comment]
Hi, thanks for your attention to our work.

I wonder which dataset and target feature(maybe `ETTh` and `OT`?) you used in your plot?
In our past experimental analysis, we also found that it is more difficult to get better prediction results on ETTh dataset, especially the prediction curve of ETTh2 is not ideal. Dataset with higher frequency may get better forecasting results, but specific features may also get bad prediction curve.

The figure below is the result of randomly selecting samples and features on the ECL dataset.
![image](https://user-images.githubusercontent.com/18379035/129828312-56371cdc-51fe-4ffa-8519-3a61a77e2e55.png)

The figure below is the result of randomly selecting samples on the ETTh1 dataset.
![image](https://user-images.githubusercontent.com/18379035/129828328-bfb225e5-72e4-4df1-a061-54b9bed72743.png)

**Comment by jkarolczak** [comment]
Thank you for the answer. The plot presents a prediction on `ETTh` (to be precise sliced part of this dataset) with `OT` as a target feature. So in general in some cases, for some datasets, it is very difficult to improve Informer predictions and you accept that it happens? Do plan any further research and improvements in the future?

**Comment by zhouhaoyi** [comment]
Generally speaking, the Informer may not be sufficient to fit all the datasets on all cases. We recommend performing better hyperparameter tunning when applying it in the practice. Moreover, as discussed in #,210, this problem may be related to the optimization loss function and you could change it in some way.

We currently working on something to improve this problem, and we hope it could help asap. 

> Thank you for the answer. The plot presents a prediction on `ETTh` (to be precise sliced part of this dataset) with `OT` as a target feature. So in general in some cases, for some datasets, it is very difficult to improve Informer predictions and you accept that it happens? Do plan any further research and improvements in the future?

**Comment by jkarolczak** [comment]
Thanks. I'm satisfied with all your answers and I think it's all for now. I'm working on a general-use application, so in the current shape, Informer doesn't fit my requirements, as I need a model which can be initially hyperparametrized in advance. It doesn't seem to be possible with Informer right now. If you will notice any significant improvements, if you would be so kind, let me know. I will be glad to take a glance and try to use it once more, as Informer seems to be a very promising model. I keep my fingers crossed for the project.

**Comment by DawenZhouCver** [comment]
这不就是白嫖伸手党吗
